src/ray/raylet_client/raylet_client.cc

```diff
@@ -296,13 +296,20 @@ Status raylet::RayletClient::FreeObjects(const std::vector<ObjectID> &object_ids
 }
 
 void raylet::RayletClient::RequestWorkerLease(
-    const TaskSpecification &resource_spec,
+    const rpc::TaskSpec &task_spec,
     const rpc::ClientCallback<rpc::RequestWorkerLeaseReply> &callback,
     const int64_t backlog_size) {
-  rpc::RequestWorkerLeaseRequest request;
-  request.mutable_resource_spec()->CopyFrom(resource_spec.GetMessage());
-  request.set_backlog_size(backlog_size);
-  grpc_client_->RequestWorkerLease(request, callback);
+  google::protobuf::Arena arena;
+  auto request =
```

Comment:
Using Arena here looks good. And if the goal is to save CPU from repeated heap allocation of sub-messages, an alternative is to reuse Message objects by just clearing them in between each usage.
https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.message_lite#MessageLite.Clear.details
If the messages can have large size variances, they can be "thrown away" after 100 uses. It can be more flexible if we also have a lot of preallocated data.

Reply:
@mwtian  Thanks for the pointing out. But here the goal is to avoid copy, right now we don't reuse replies around.